docs/style-guide.md

```diff
@@ -0,0 +1,439 @@
+  * This has a lot of good points and is well worth a read, but be warned that some of the details about implementation are somehwat out of date (2017). It has a video too.
+* [Good Practices for Writing Rust Libraries](https://pascalhertleif.de/artikel/good-practices-for-writing-rust-libraries/)
+  * Shorter and more focussed on the act of coding the libraries, and less about API design. Also potentially out of date in places.
+* [Strategies for Returning References in Rust](http://bryce.fisher-fleig.org/blog/strategies-for-returning-references-in-rust/index.html)
+  * Though I don't believe we should be doing this, it's still an interesting read.
+
+## Other Useful Links
+
+* Write and run Rust snippets: https://play.rust-lang.org
+  * You can save snippets in permantent links and incluce them as working examples in docs.
+* Write code and see what it compiles to: https://rust.godbolt.org
+  * Note that you need to enable compiler optimizations via `-C opt-level=3` if you are looking to meaningfully compare two code snippets.
+
+# Naming Conventions
+
+### No Special Naming :: required
+
+Follow the naming advice in [Naming - Rust API Guidelines](https://rust-lang.github.io/api-guidelines/naming.html) at all times, and use [Rust RFC 0430](https://github.com/rust-lang/rfcs/blob/master/text/0430-finalizing-naming-conventions.md) for any ambiguous issues (e.g. Crate naming).
+
+Based on [this comment](https://github.com/rust-lang/api-guidelines/issues/29#issuecomment-342745898) I think the advice for Crate names should be:
+
+* Infer Crate names from package names where feasible:
+  * A package is the directory in which the Cargo.toml exists.
+* Prefer single (lowercase) words for package names where feasible
+* Use __lowercase with hyphen__ to split words in package names if needed
+
+This means that while Crate names are not trivially 
+I believe this would result in Crate names (which must be Rust identifiers) being identical to the package name at all times.
+
+## Module Layout
+
+Modules form the basic units of imported code in Rust, so identifying where to split functionality and what to call modules is essential. I suspect this is the right level to divide things like collation vs segmentation etc. but it's also far too early to make hard recommendations about this, so I proposed to leave this issue open for the time being (moving code between modules later should be relatively easy).
+
+### Don't have "mixed" Crates :: suggested
+
+The current thinking is that you should either:
+
+* Use a single `lib.rs` file for a Crate with everything in it
+* Use multiple Rust sources and then only use `lib.rs` to export modules
+
+Which of these options you chose is dependent on the size/complexity of the module(s).
+
+There might be cases where an intermediate "mixed" Crate is worthwhile, but it should be clearly documented as to why it is necessary.
+
+## Shared Internal Code
```

Comment:
Nit: I am not clear what type of section this is:
- statement that n-p s-modules exist
- suggestion to use non-public sub-modules for shared code

Reply:
Agree. It reads like a statement, not a guideline.

Reply:
It is a statement, it's the preface to the section below, although that ended up with just one guideline. I think this is going back the the roots of this doc as more of an introduction to people who don't know much Rust. Happy to remove or reword it.